### PR TITLE
feat(thes-2530): starting Ceres as daemon and retry deferred QRPs

### PR DESCRIPTION
**Pull request by anbonifacio**
## Description

Please write here a summary of the change(s) in this PR.

### Checklist

I confirm that before submitting this PR I have:

- [x] successfully run `cargo build` on my local environment
- [x] formatted all Rust files with `cargo fmt`
- [x] checked all code hints with `cargo clippy`
- [x] successfully run `cargo test` on my local environment
- [ ] successfully run automatic tests in QA (namespace `xx-yyy`)
- [ ] successfully run manual tests in QA (namespace `xx-yyy`)
- [ ] updated the project README with new relevant information
- [ ] added new documentation for Backstage
    - [ ] updated resources (`catalog-info.yaml`, `providesApis`, `consumesApi`, `mkdocs.yml`) with proper information
    - [x] updated file(s) under `docs/`
        - [x] with a comprehensive description of the new features
        - [x] with useful development and testing commands
- [ ] updated Confluence page with new relevant information

### CODE REVIEW

File: src/bin/http_server.rs [hunk 1]
```diff
@@ -37,6 +36,36 @@ async fn main() -> IoResult<()> {
...
+
+async fn run_daemon(db: Db, cerved_qrp_client: CervedQrpClient, s3client: S3Client) -> IoResult<()> {
+    info!("Ceres C'è: starting as Daemon");
+    let deferred_qrp_service = DeferredQrpService {
+        db: db.clone(),
+        qrp_client: cerved_qrp_client.clone(),
+        s3_client: s3client.clone(),
+    };
+
+    let interval = tokio::time::interval(std::time::Duration::from_secs(10));
```
Comment (anbonifacio) [inline src/bin/http_server.rs:53]: Quale potrebbe essere un buon intervallo?

File: src/persistence/mod.rs [hunk 2]
```diff
@@ -72,12 +73,35 @@ pub struct Db {
...
             .get_result(&mut self.connection.get()?)?)
     }
+
+    pub fn find_deferred_qrp_entries(&self) -> Result<Vec<QrpSelectEntry>> {
+        use schema::qrp_entry::dsl::*;
+
+        Ok(qrp_entry
+            .select(QrpSelectEntry::as_select())
+            .filter(status.eq(Deferred))
+            .limit(10)
```
Comment (anbonifacio) [inline src/persistence/mod.rs:91]: Should we limit the number of deferred entries to process?

File: src/bin/http_server.rs [hunk 3]
```diff
@@ -50,16 +52,15 @@ async fn run_daemon(db: Db, cerved_qrp_client: CervedQrpClient, s3client: S3Clie
...
     };
 
-    let interval = tokio::time::interval(std::time::Duration::from_secs(10));
-
-    let forever = stream::unfold(interval, |mut interval| async {
-        interval.tick().await;
-        let res = deferred_qrp_service.process_deferred_qrp().await;
-        Some((res, interval))
+    let forever = task::spawn(async move {
+        let mut interval = tokio::time::interval(std::time::Duration::from_secs(FIFTEEN_MINUTES_IN_SECONDS));
```
Comment (albx79) [inline src/bin/http_server.rs:56]: you don't need to spawn anything. You can just write a plain loop directly in the async function. The executor takes care of running it a bit at a time and never blocks (provided you have some `.await` inside the loop)

File: src/bin/http_server.rs [hunk 4]
```diff
@@ -37,6 +37,31 @@ async fn main() -> IoResult<()> {
...
+    let deferred_qrp_service = DeferredQrpService {
+        db: db.clone(),
+        qrp_client: cerved_qrp_client.clone(),
+        s3_client: s3client.clone(),
+    };
+
+    let mut interval = tokio::time::interval(std::time::Duration::from_secs(FIFTEEN_MINUTES_IN_SECONDS));
+    loop {
+        interval.tick().await;
+        let _ = deferred_qrp_service.process_deferred_qrp().await;
```
Comment (albx79) [inline src/bin/http_server.rs:57]: See, I always find one last thing.

`process_deferred_qrp` returns a result. By discarding it, you're doing the rust equivalent of swallowing an exception in java.
```suggestion
        deferred_qrp_service.process_deferred_qrp().await?;
```

Comment (anbonifacio) [reply src/bin/http_server.rs:57]: Ok now I see why I took the IDE suggestion to add `let _ =`: the function returns an `IoResult`, I have to convert the `anyhow` into it in some way. Any idea?

Comment (anbonifacio) [reply src/bin/http_server.rs]: I've resolved using `anyhow::Result` as return typw for the main fn, and I mapped the error from actix `HttpServer::new`

**Review by albx79 — APPROVED** [review]

**Pull request merged**